doc/bgp_loading_optimization/bgp-loading-optimization-hld.md

```diff
@@ -0,0 +1,421 @@
+5. `orchagent` injects its processed routing message into `ASIC_DB` with `sairedis` APIs 
+6. `syncd` gets notified of the newly added routing message in `ASIC_DB` due to the subscription
+7. `syncd` processes the new routing message, then invokes `SAI` APIs to finally inject the new routing message into the hardware
+
+
+**NOTE**: This is not the [whole workflow for routing](https://github.com/SONiC-net/SONiC/wiki/Architecture#routing-state-interactions), we ignore the Linux kernel part since we currently focus only on the SONiC part.
+
+<figure align="center">
+    <img src="images/sonic-workflow.png" width="60%" height=auto>
+    <figcaption>Figure 2. SONiC BGP loading workflow</figcaption>
+</figure>
+
+
+### Single-threaded orchagent
+
+Figure 3 explains how `orchagent` transfers routing data from `APPL_DB` to `ASIC_DB`.
+
+`RouteOrch`, as a component of `orchagent`, has its `ConsumerStateTable` subscribed to `ROUTE_TABLE_CHANNEL` event. With this subscription, whenever `fpmsyncd` injects new routing data into `APPL_DB`, `orchagent` gets notified. Once notified, `orchagent` handles the following 3 tasks in serial.
+
+1. use `pops` to fetch new routes from `APPL_DB`:
+     - pop prefix from ROUTE_TABLE_SET 
+     - traverse these prefixes and retrieve the temporary key data of _ROUTE_TABLE corresponding to the prefix
+     - set key in ROUTE_TABLE 
+     - delete temporary key in _ROUTE_TABLE
+2. call `addToSync` to record the new routes to a local file `swss.rec`
+3. call `doTask` to parse new routes one by one and store the processed data in the EntityBulker, and flush the data in EntityBulker to ASIC_DB as a whole
+
+
+The main performance bottleneck here lies in the linearity of the 3 tasks.
+
+<br>
+
+<figure align=center>
+    <img src="images/orchagent-workflow.png" width="60%" height=auto>
+    <figcaption>Figure 3. Orchagent workflow<figcaption>
+</figure>  
+
+
+### Single-threaded syncd
+
+`syncd` shares the similar problem (job linearity) with `orchagent`, the only difference is that `syncd` moves information from `ASIC_DB` to the hardware. 
+
+<br>
+
+<figure align=center>
+    <img src="images/syncd-workflow.jpg" width="60%" height=auto>
+    <figcaption>Figure 4. Syncd workflow<figcaption>
+</figure>  
+
+
+### Redundant APPL_DB I/O traffic
+
+There is much Redis I/O traffic during the BGP loading process, from which we find two sources of unnecessary traffic.
+
+#### fpmsyncd flushes on every incoming route
+In the original design, `fpmsyncd` maintains a variable `pipeline`. Each time `fpmsyncd` receives a route from `zebra`, it processes the route and puts it in the `pipeline`. Every time the `pipeline` receives a route, it flushes the route to `APPL_DB`. If the size of the incoming route exceeds the size of the `pipeline` itself, the `pipeline` performs multiple flushes to make sure the received routes are written into `APPL_DB` completely. 
```

Comment:
>Every time the `pipeline` receives a route, it flushes the route to `APPL_DB` [](http://example.com/codeflow?start=174&length=77)

I do not believe fpmsyncd flushes APPL_DB on every route. fpmsyncd uses m_routeTable and a `pipeline` with size=128. So normally it will flush every 128 routes.

Reply:
Sorry for the bad wording, by "per route", we actually refer to "per batch" (each time the Lua script is executed)